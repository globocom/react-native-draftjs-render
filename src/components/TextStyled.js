/*
 * Copyright (c) 2017, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

// @flow

import React from 'react';
import {
  Text,
  StyleSheet,
} from 'react-native';

import type { TextStyledPropsType } from './types';

const styles = StyleSheet.flatten({
  bold: {
    fontWeight: 'bold',
  },
  italic: {
    fontStyle: 'italic',
  },
  link: {
    textDecorationLine: 'underline',
  },
  underline: {
    textDecorationLine: 'underline',
  },
  strikethrough: {
    textDecorationLine: 'line-through',
  },
});

const getStyles = (itemType: any, customStyles?: Object): any => {
  if (!customStyles) return [styles[itemType]];
  if (typeof itemType === 'string') return [styles[itemType], customStyles[itemType]];

  const defaultTextStyles = {};
  const customTextStyles = {};
  itemType.forEach((i: string) => {
    Object.assign(defaultTextStyles, styles[i]);
    if (customStyles) Object.assign(customTextStyles, customStyles[i]);
  });
  const newStyles = [defaultTextStyles, customTextStyles];
  return newStyles;
};

const TextStyled = (props: TextStyledPropsType): any => {
  const textStyle = getStyles(props.type, props.customStyles);

  if (props.onPress) {
    return <Text style={textStyle} onPress={props.onPress}>{props.text}</Text>;
  }
  return <Text style={textStyle}>{props.text}</Text>;
};

TextStyled.defaultProps = {
  text: '',
  onPress: undefined,
};

export default TextStyled;
