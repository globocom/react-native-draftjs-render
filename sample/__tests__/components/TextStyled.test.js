/*
 * Copyright (c) 2017, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

/* eslint-env jest */

import 'react-native';

import React from 'react';
import renderer from 'react-test-renderer';

import TextStyled from '../../../src/components/TextStyled';

it('renders correctly without onPress', () => {
  const tree = renderer.create(<TextStyled
    type="paragraph"
  />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with onPress', () => {
  const tree = renderer.create(<TextStyled
    type="paragraph"
    onPress={() => null}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with customStyles and multiple types', () => {
  const customStyles = {
    bold: {
      fontWeight: 'bold',
    },
    italic: {
      fontStyle: 'italic',
    },
    link: {
      textDecorationLine: 'underline',
    },
    underline: {
      textDecorationLine: 'underline',
    },
    strikethrough: {
      textDecorationLine: 'line-through',
    },
  };
  const typeArray = ['bold', 'italic'];
  const tree = renderer.create(<TextStyled
    type={typeArray}
    customStyles={customStyles}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with customStyles and single type', () => {
  const customStyles = {
    bold: {
      fontWeight: 'bold',
    },
  };
  const type = 'bold';
  const tree = renderer.create(<TextStyled
    type={type}
    customStyles={customStyles}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});
