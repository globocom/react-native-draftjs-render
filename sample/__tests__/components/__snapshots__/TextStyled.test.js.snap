// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with customStyles and multiple types 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "fontStyle": "italic",
        "fontWeight": "bold",
      },
      Object {
        "fontStyle": "italic",
        "fontWeight": "bold",
      },
    ]
  }
>
  
</Text>
`;

exports[`renders correctly with customStyles and single type 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "fontWeight": "bold",
      },
      Object {
        "fontWeight": "bold",
      },
    ]
  }
>
  
</Text>
`;

exports[`renders correctly with onPress 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
  onPress={[Function]}
  style={
    Array [
      undefined,
    ]
  }
>
  
</Text>
`;

exports[`renders correctly without onPress 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
  style={
    Array [
      undefined,
    ]
  }
>
  
</Text>
`;
